### PR TITLE
table: support check constraint in insert and update

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
close https://github.com/pingcap/tidb/issues/17425

### What is changed and how it works?

- Add `Constraint` class
- Build constraint object while initilize table object
- Eval and check constraint expression  in `UpdateRecord` and `AddRecord`

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression

Related changes

 - None

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -148,17 +167,19 @@ func TableFromMeta(allocs autoid.Allocators, tblInfo *model.TableInfo) (table.Ta
...
 
 // initTableCommon initializes a TableCommon struct.
-func initTableCommon(t *TableCommon, tblInfo *model.TableInfo, physicalTableID int64, cols []*table.Column, allocs autoid.Allocators) {
+func initTableCommon(t *TableCommon, tblInfo *model.TableInfo, physicalTableID int64, cols []*table.Column, constraints []*table.Constraint, allocs autoid.Allocators) {
 	t.tableID = tblInfo.ID
 	t.physicalTableID = physicalTableID
 	t.allocs = allocs
 	t.meta = tblInfo
 	t.Columns = cols
+	t.PublicConstraints = constraints
```
Comment (AilinKid) [inline table/tables/tables.go]: how about rename `t.PublicConstraints` to `t.Constraints` ,  which means the constraints in all state?

File: table/tables/tables.go [hunk 2]
```diff
@@ -284,6 +305,23 @@ func (t *TableCommon) WritableCols() []*table.Column {
...
+func (t *TableCommon) WritableConstraint() []*table.Constraint {
+	if len(t.WritableConstraints) > 0 {
+		return t.WritableConstraints
+	}
+	if t.PublicConstraints == nil {
+		return nil
+	}
+	writeableConstraint := make([]*table.Constraint, 0, len(t.PublicConstraints))
+	for _, con := range t.PublicConstraints {
+		if con.State == model.StateDeleteOnly || con.State == model.StateDeleteReorganization {
```
Comment (AilinKid) [inline table/tables/tables.go:317]: constraint only has 3 state: None -> WriteOnly -> Public, but the code correctness here is ok.

File: table/tables/tables.go [hunk 3]
```diff
@@ -352,13 +392,26 @@ func (t *TableCommon) UpdateRecord(ctx sessionctx.Context, h int64, oldData, new
...
 		if shouldWriteBinlog(ctx) && !t.canSkipUpdateBinlog(col, value) {
 			binlogColIDs = append(binlogColIDs, col.ID)
 			binlogOldRow = append(binlogOldRow, oldData[col.Offset])
 			binlogNewRow = append(binlogNewRow, value)
 		}
 	}
 
+	if len(t.WritableConstraint()) != 0 {
+		for _, constraint := range t.WritableConstraint() {
+			ok, isNull, err := constraint.ConstraintExpr.EvalInt(ctx, chunk.MutRowFromDatums(rowToCheck).ToRow())
```
Comment (AilinKid) [inline table/tables/tables.go:405]: should it be EvalBool?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: table/tables/tables.go (hunk 3 above)
Comment (imtbkcat) [reply table/tables/tables.go:405]: `EvalBool` has some problem, which sometimes can not return correct `isNull`

Comment (AilinKid) [reply table/tables/tables.go:405]: ok

**Review by AilinKid — COMMENTED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**